depthai_ros_driver/include/depthai_ros_driver/impl/ros_agnostic.hpp

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef DEPTHAI_ROS_DRIVER__IMPL__ROS_AGNOSTIC_HPP
+#define DEPTHAI_ROS_DRIVER__IMPL__ROS_AGNOSTIC_HPP
+
+#include <depthai_ros_driver/ros_agnostic.hpp>
+
+namespace rr {
+namespace ros_agnostic {
+
+/*
+ * Here we will keep all details implemntaion of within ros_agnostic ns. This
+ * includes: Publisher, Subscription, Server, Timer...
+ */
+
+//==============================================================================
+class Publisher
+{
+public:
+  /// @brief Publish ros msg
+  template<typename Msg>
+  void publish(Msg& msg)
+  {
+    #if defined(USE_ROS2)
+    // move the ptr and subscriber check here
+    auto pub = std::static_pointer_cast<rclcpp::Publisher<Msg>>(_ptr);
+    if (pub && pub->get_subscription_count() > 0)
+      pub->publish(msg);
+    #else
+    if (_pub.getNumSubscribers() > 0);
+    _pub.publish(msg);
+    #endif
+  }
+
+private:
+  /// @brief Create ros publisher
+  template<class Msg>
+  void create_publisher(
```

Comment:
Should this be static?

Reply:
If a factory method of `static make_publisher()` is preferred, this can be done.